include/rocksdb/utilities/transaction.h

```diff
@@ -227,6 +237,53 @@ class Transaction {
   // transaction before Commit.
   virtual Status Commit() = 0;
 
+  // In addition to Commit(), also creates a snapshot of the db after all
+  // writes by this txn are visible to other readers.
+  // Currently only supported by WriteCommittedTxn. Calling this method on
+  // other types of transactions will return non-ok Status resulting from
+  // Commit() or a `NotSupported` error.
+  // notifier will be notified upon next snapshot creation. Nullable.
+  // ret non-null output argument storing a shared_ptr to the newly created
+  // snapshot.
+  Status CommitAndCreateSnapshot(
+      std::shared_ptr<TransactionNotifier> notifier =
+          std::shared_ptr<TransactionNotifier>(),
+      TxnTimestamp ts = kMaxTxnTimestamp,
+      std::shared_ptr<const Snapshot>* snapshot = nullptr) {
+    TxnTimestamp commit_ts = GetCommitTimestamp();
+    if (commit_ts == kMaxTxnTimestamp) {
+      if (ts == kMaxTxnTimestamp) {
+        return Status::InvalidArgument("Commit timestamp unset");
+      } else {
+        const Status s = SetCommitTimestamp(ts);
+        if (!s.ok()) {
+          return s;
+        }
+      }
+    } else if (ts != kMaxTxnTimestamp) {
+      if (ts != commit_ts) {
+        // For now we treat this as error.
+        return Status::InvalidArgument("Different commit ts specified");
+      }
+    }
+    // Do not try to dereference prev_snapshot because it can be destroyed.
+    const Snapshot* const prev_snapshot = GetSnapshot();
+    SetSnapshotOnNextOperation(notifier);
+    Status s = Commit();
+    if (!s.ok()) {
+      return s;
+    }
+    std::shared_ptr<const Snapshot> new_snapshot = GetSharedSnapshot();
+    if (new_snapshot.get() == prev_snapshot) {
+      return Status::NotSupported();
+    }
+
+    if (snapshot) {
+      *snapshot = new_snapshot;
+    }
+    return s;
```

Comment:
Does the implementation need to stay in public header? Can it be moved into a .cc file?